oeps/oep-0018-bp-python-dependencies.rst

```diff
@@ -0,0 +1,402 @@
+we need a separate set of requirements files which list every single package
+needed for each usage context, with exact versions of each for reproducible
+test runs and consistent development and production environments.  We can
+generate these automatically using `pip-tools`_, which consists of two related
+utilities:
+
+* ``pip-compile`` generates a requirements file from one or more high-level
+  input requirements files, listing exact versions of every listed and
+  indirect dependency needed to satisfy the given constraints.
+* ``pip-sync`` ensures that the current virtualenv contains exactly (and only)
+  the packages listed in the given requirements files, installing, upgrading,
+  and uninstalling packages as needed.
+
+Open edX packages should use an ``upgrade`` make target to use ``pip-compile``
+to automatically update the detailed requirements files
+(``requirements/*.txt``) to use the newest available packages which satisfy
+the constraints in the direct dependencies files.  These generated files are
+then used anywhere that runs a command to install dependencies: ``tox.ini``,
+``.travis.yml``, the ``requirements`` make target (for updating a local
+development environment), etc.
+
+Sometimes ``pip-compile`` will be unable to find a suitable version of a
+dependency for the output file because there are incompatible version
+constraints in the input files and/or the stated installation requirements
+of the other dependencies.  In cases like this, installing and running
+`pipdeptree`_ can help identify the conflicting constraints so at least one
+of them can be sufficiently relaxed such that a version of the dependency
+exists which satisfies them all.
+
+.. _pip-tools: https://github.com/jazzband/pip-tools
+.. _pipdeptree: https://github.com/naiquevin/pipdeptree
+
+Automate Updates of Exact Dependency Specifications
+---------------------------------------------------
+
+While we want all dependencies explicitly pinned in order to benefit from
+consistent testing and development environments, it isn't acceptable to leave
+these versions untouched for long stretches of time.  Packages we depend on
+routinely release new versions to address security issues, fix bugs, and add
+new features.  While we don't necessarily need to update our repositories
+every time a new dependency version is released, we do want to keep them
+current enough that upgrading a single package to fix a known issue doesn't
+require suddenly adapting to a few years' worth of API changes that we didn't
+pay attention to.
+
+Each Open edX repository should have the following:
+
+* An ``upgrade`` make target as described above, to update the pinned versions
+  of all dependencies (and account for any new or removed indirect
+  dependencies).
+* An automated test suite with reasonably good code coverage, configured to
+  be run on new GitHub pull requests.
+* A service configured to periodically auto-generate a GitHub pull request
```

Comment:
Am I understanding this correctly: Until the next time this service runs, the automatically generated requirements `.txt` file may not match the output of the `.in` file, right?

Reply:
Anytime a developer updates one of the `.in` files, they should run `make upgrade` to update all of the `.txt` files.  The periodic auto-generated pull requests (probably weekly or so) are just to notify us of updates to packages that we aren't deliberately holding back, so we don't fall too far behind.  Those auto-generated PRs would only update the `.txt` files, since the input constraints won't have changed (and the suggested changes still need to be reviewed and deliberately merged).